<a href="https://colab.research.google.com/github/Gizeh-Gutierrez/Maestria_programacion/blob/main/Actividad_Calificada_07_Datos_Duplicados_271752.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos duplicados**
### Detección, cuantificación y tratamiento de duplicados en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: duplicados |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Hotel Booking Demand (Antonio, Almeida y Nunes, 2019) |
| **Fecha de entrega** | 25 Septiembre 2026|

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo** |Gizeh Anaid Gutierrez Lopez |
| **Matrícula** |271752 |

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en los resultados que obtuviste (cita cifras).
7. Usa los métodos vistos en el notebook de la sesión: `duplicated()`, `drop_duplicates()`, `value_counts()`, `sort_values()`, `groupby()`, `pd.concat()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Duplicados_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset **Hotel Booking Demand** contiene 119,390 reservas de dos hoteles de Portugal (un hotel urbano, *City Hotel*, y un hotel vacacional, *Resort Hotel*) con llegada entre julio de 2015 y agosto de 2017. Cada fila es una reserva y hay 32 columnas. Algunas relevantes para esta actividad:

| Columna | Descripción |
|---|---|
| `hotel` | `City Hotel` o `Resort Hotel` |
| `is_canceled` | 1 si la reserva se canceló, 0 si no |
| `lead_time` | Días entre la reserva y la fecha de llegada |
| `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month` | Fecha de llegada |
| `adults`, `children`, `babies` | Número de huéspedes |
| `country` | País de origen (código ISO de 3 letras) |
| `market_segment` | Segmento de mercado (`Online TA`, `Groups`, `Direct`, …) |
| `deposit_type` | Tipo de depósito (`No Deposit`, `Non Refund`, `Refundable`) |
| `agent`, `company` | Identificador de la agencia o empresa que hizo la reserva (con faltantes) |
| `customer_type` | Tipo de cliente (`Transient`, `Group`, …) |
| `adr` | Tarifa diaria promedio (*average daily rate*) |
| `reservation_status` | Estado final: `Check-Out`, `Canceled`, `No-Show` |
| `reservation_status_date` | Fecha del último cambio de estado |

> **Importante**: el dataset **no incluye un identificador de reserva**. Tenlo presente al interpretar los duplicados.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Diagnóstico inicial** | 20 | 18–20: conteos y porcentajes correctos con los tres valores de `keep`; duplicados por hotel calculados; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o en la conclusión. | 1–13: conteos incompletos o con errores; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Inspección y criterio** | 25 | 23–25: grupos inspeccionados y ordenados; tamaños de grupo analizados; comparación duplicados vs. total; argumento sólido sobre si son errores, con evidencia. | 18–22: análisis correcto pero con argumento parcial o poca evidencia. | 1–17: inspección superficial; argumento sin datos. | 0: no se entregó o es incorrecto. |
| **3. Impacto en el análisis** | 25 | 23–25: métricas por hotel antes y después con `groupby()`; tabla comparativa con `pd.concat()`; interpretación del efecto en el análisis y en un modelo predictivo. | 18–22: métricas correctas; tabla o interpretación incompletas. | 1–17: métricas con errores o sin comparación. | 0: no se entregó o es incorrecto. |
| **4. Tratamiento y verificación** | 20 | 18–20: limpieza verificada; duplicados parciales detectados e interpretados; registro de decisión completo. | 14–17: limpieza correcta; análisis parcial o registro incompleto. | 1–13: limpieza sin verificación o análisis parcial erróneo. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.2 y 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [1]:
import pandas as pd

# Carga del dataset Hotel Booking Demand desde GitHub (TidyTuesday, 2020-02-11)
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-02-11/hotels.csv'
hoteles = pd.read_csv(url)

# Dimensiones y primeras filas
print('Dimensiones:', hoteles.shape)
hoteles.head()

Dimensiones: (119390, 32)


,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,304.0,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,240.0,NaN,0,Transient,98.0,0,1,Check-Out,2015-07-03


---
## Actividad 1. Diagnóstico inicial (20 puntos)

### 1.1 Conteo de duplicados exactos (5 pts)

Calcula el número de filas marcadas como duplicadas en `hoteles` considerando **todas** las columnas, con `keep='first'`, `keep='last'` y `keep=False`. Imprime los tres resultados con un texto que explique qué significa cada uno.

In [2]:
# Conteo de duplicados.

dup_first = hoteles.duplicated(keep='first').sum()
dup_last = hoteles.duplicated(keep='last').sum()
dup_false = hoteles.duplicated(keep=False).sum()

print(f"Duplicados keep='first': {dup_first}")

print(f"Duplicados keep='last': {dup_last}")

print(f"Duplicados keep=False: {dup_false}")


Duplicados keep='first': 31994
Duplicados keep='last': 31994
Duplicados keep=False: 40165


El data set tiene un total de  119,390 reservas
los duplicados  que se tienen son son los siguientes:
Duplicados con keep='first': 31,994

*   Duplicados con keep='first':**31,994**
*   Duplicados con keep='last': **31,994**
*   Duplicados con keep=False: **40,165**


El hecho de que los valores de keep='first' y keep='last' sean iguales indica que la cantidad de registros repetidos es la misma independientemente de si se conserva la primera o la última aparición de cada grupo de filas idénticas.

El resultado de 40,165 filas con keep=False significa que ese número de registros pertenece a algún grupo de duplicados. Es decir, aproximadamente el 33.6% de las observaciones del conjunto de datos forman parte de grupos de filas completamente idénticas:

(40165/119390)*100= 33.6%

---



La diferencia entre 40,165 y 31,994 se debe a que keep=False cuenta todas las filas involucradas en duplicados, mientras que keep='first' y keep='last' excluyen una ocurrencia de cada grupo.

Debido a que no existe un identificador único de reserva, no puede afirmarse  que se trate de errores o duplicados reales. Es posible que algunas reservas distintas tengan exactamente los mismos valores en las 32 variables registradas.

### 1.2 Magnitud del problema (5 pts)

Con los resultados de 1.1, calcula e imprime:

- El **porcentaje** de filas sobrantes respecto al total.
- El número de **grupos repetidos** (combinaciones de valores que aparecen más de una vez).

In [3]:
# Filas sobrantes
filas_sobrantes = hoteles.duplicated(keep='first').sum()

# Porcentaje
porcentaje_sobrantes = (filas_sobrantes / len(hoteles)) * 100

# Grupos repetidos
grupos_repetidos = (
    hoteles.groupby(list(hoteles.columns))
           .size()
           .gt(1)
           .sum()
)

print(f"Porcentaje de filas sobrantes: {porcentaje_sobrantes:.2f}%")
print(f"Grupos repetidos: {grupos_repetidos}")

Porcentaje de filas sobrantes: 26.80%
Grupos repetidos: 17


El **26.80%** de las filas son registros repetidos respecto a una observación idéntica previa. En términos prácticos, esto significa que 31,994 de las 119,390 filas podrían eliminarse si se conserva una sola ocurrencia de cada registro duplicado.

Se encontraron **17 grupos** repetidos, por lo que solo 17 combinaciones de valores distintas aparecen más de una vez. Esto indica que los duplicados no están distribuidos en muchos patrones diferentes.


### 1.3 Duplicados por hotel (5 pts)

Determina **cuántas** filas sobrantes (`keep='first'`) hay en cada hotel y qué **porcentaje** representan respecto al total de reservas de ese hotel.

Sugerencia: crea una copia de `hoteles`, agrega una columna booleana `es_duplicado` con el resultado de `duplicated()` y agrupa por `hotel`.

In [4]:
# Copia
hoteles_dup = hoteles.copy()

# Booleana de duplicados
hoteles_dup['es_duplicado'] = hoteles_dup.duplicated(keep='first')

# Datos por Hotel
resumen = hoteles_dup.groupby('hotel').agg(
    total_reservas=('hotel', 'size'),
    filas_sobrantes=('es_duplicado', 'sum')
)

resumen['porcentaje_sobrantes'] = (
    resumen['filas_sobrantes'] / resumen['total_reservas'] * 100
)

print(resumen)


              total_reservas  filas_sobrantes  porcentaje_sobrantes
hotel                                                              
City Hotel             79330            25902             32.650952
Resort Hotel           40060             6092             15.207189


El City Hotel tiene **25,902** filas sobrantes 32.65%, mientras que el Resort Hotel tiene 6,092 filas sobrantes 15.21%.
Esto indica que City Hotel tiene mas registros repedidos.

### 1.4 Conclusión (5 pts)

¿Qué tan grave es el problema de duplicados en este dataset? ¿Afecta por igual a ambos hoteles? Responde con cifras.

**Respuesta 1.4:**
El problema de duplicados es grave en este dataset.

Se tienen 31,994 filas sobrantes de un total de 119,390 registros.
Esto representa el 26.80% del dataset, es decir, más de 1 de cada 4 filas corresponde a una repetición.

El problema no afecta por igual a ambos hoteles:

City Hotel: 25,902 filas sobrantes 32.65%.
Resort Hotel: 6,092 filas sobrantes 15.21%.

Por lo tanto, la proporción de duplicados en el City Hotel es más del doble que en el Resort Hotel. La mayoría de los registros repetidos se concentra en el City Hotel.

Sin embargo, como se comento anteriormente  el dataset no contiene un identificador único, por lo que estos registros idénticos no pueden considerarse automáticamente errores.

---
## Actividad 2. Inspección y criterio de duplicado (25 puntos)

### 2.1 Ver los grupos repetidos (6 pts)

Filtra las filas involucradas en duplicados (`keep=False`), **ordénalas** con `sort_values()` para que las copias queden juntas y muestra las primeras 12 filas con estas columnas:

`hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adults`, `country`, `market_segment`, `deposit_type`, `adr`, `reservation_status`.

Elige tú las columnas de ordenamiento y explica tu elección en un comentario.

In [5]:

duplicados = hoteles[hoteles.duplicated(keep=False)]

# Ordenamiento elegido:
# Se ordena por las variables que describen la reserva para que las filas idénticas queden juntas y
# sea más fácil identificar grupo duplicado.
duplicados = duplicados.sort_values(
    by=[
        'hotel',
        'lead_time',
        'arrival_date_year',
        'arrival_date_month',
        'arrival_date_day_of_month',
        'adults',
        'country',
        'market_segment',
        'deposit_type',
        'adr',
        'reservation_status'
    ]
)

# Mostrar columnas
columnas = [
    'hotel',
    'lead_time',
    'arrival_date_year',
    'arrival_date_month',
    'arrival_date_day_of_month',
    'adults',
    'country',
    'market_segment',
    'deposit_type',
    'adr',
    'reservation_status'
]

print(duplicados[columnas].head(12))

            hotel  lead_time  arrival_date_year arrival_date_month  \
40643  City Hotel          0               2015             August   
40644  City Hotel          0               2015             August   
40772  City Hotel          0               2015             August   
40802  City Hotel          0               2015             August   
40821  City Hotel          0               2015             August   
40838  City Hotel          0               2015             August   
76792  City Hotel          0               2015             August   
76793  City Hotel          0               2015             August   
76794  City Hotel          0               2015             August   
41067  City Hotel          0               2015             August   
41073  City Hotel          0               2015             August   
41071  City Hotel          0               2015             August   

       arrival_date_day_of_month  adults country market_segment deposit_type  \
40643    

Se realizo el orden por hotel, fecha de llegada, lead_time, número de adultos, país, segmento de mercado, tipo de depósito, tarifa (adr) y estado de la reserva, porque estas variables describen las características de una reserva. Al utilizar estas columnas en sort_values(), las filas con valores idénticos quedan agrupadas una junto a la otra, haciendo mas facil identificar las repetidas.

### 2.2 Tamaño de los grupos (6 pts)

Usa `value_counts()` sobre el `DataFrame` completo para contar cuántas veces aparece cada combinación de filas. Considera que las columnas `agent` y `company` tienen faltantes (revisa el parámetro `dropna`).

Con ese resultado:

1. Muestra **cuántos grupos** hay de cada tamaño (2 copias, 3 copias, …). Basta con los 10 tamaños más frecuentes.
2. Indica el **tamaño del grupo más grande** y muestra su combinación de valores.

In [6]:
conteos = hoteles.value_counts(dropna=False)

# Numero de grupos
tam_grupos = conteos.value_counts()

print("Los 10 tamaños de grupo  frecuentes:")
print(tam_grupos.head(10))

# Grupo más grande
max_tam = conteos.max()

print("\nTamaño del grupo más grande:", max_tam)

# Combinación de valores del grupo más grande
grupo_mayor = conteos[conteos == max_tam]
print("\nCombinación de valores del grupo más grande:")
print(grupo_mayor)

Los 10 tamaños de grupo  frecuentes:
count
1     79225
2      5034
3       969
4       497
5       275
6       171
10      123
7       115
8       103
20       97
Name: count, dtype: int64

Tamaño del grupo más grande: 180

Combinación de valores del grupo más grande:
hotel       is_canceled  lead_time  arrival_date_year  arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  meal  country  market_segment  distribution_channel  is_repeated_guest  previous_cancellations  previous_bookings_not_canceled  reserved_room_type  assigned_room_type  booking_changes  deposit_type  agent  company  days_in_waiting_list  customer_type  adr    required_car_parking_spaces  total_of_special_requests  reservation_status  reservation_status_date
City Hotel  1            277        2016               November            46                        7                          1                        2                 

Tamaño del grupo más grande es de 180 registros idénticos lo que sugiere que los duplicados están concentrados en algunas combinaciones.

### 2.3 ¿En qué se distinguen las filas duplicadas? (6 pts)

Compara las filas involucradas en duplicados (`keep=False`) con el dataset completo en:

- Distribución de `market_segment` (proporciones).
- Distribución de `deposit_type` (proporciones).
- Tasa de cancelación (`is_canceled`).

In [7]:
duplicados = hoteles[hoteles.duplicated(keep=False)]

# Market segment
print("Market Segment")
print("Dataset completo")
print(hoteles['market_segment'].value_counts(normalize=True))

print("\nDuplicados")
print(duplicados['market_segment'].value_counts(normalize=True))

# Deposit type
print("\nDeposit Type")
print("Dataset completo")
print(hoteles['deposit_type'].value_counts(normalize=True))

print("\nDuplicados")
print(duplicados['deposit_type'].value_counts(normalize=True))

# Tasa de cancelación
print("\nTasa de cancelación")
print("Dataset completo:", hoteles['is_canceled'].mean())
print("Duplicados:", duplicados['is_canceled'].mean())

Market Segment
Dataset completo
market_segment
Online TA        0.473046
Offline TA/TO    0.202856
Groups           0.165935
Direct           0.105587
Corporate        0.044350
Complementary    0.006223
Aviation         0.001985
Undefined        0.000017
Name: proportion, dtype: float64

Duplicados
market_segment
Groups           0.416109
Offline TA/TO    0.303996
Online TA        0.205876
Corporate        0.036699
Direct           0.034981
Complementary    0.001867
Aviation         0.000473
Name: proportion, dtype: float64

Deposit Type
Dataset completo
deposit_type
No Deposit    0.876464
Non Refund    0.122179
Refundable    0.001357
Name: proportion, dtype: float64

Duplicados
deposit_type
No Deposit    0.641678
Non Refund    0.356430
Refundable    0.001892
Name: proportion, dtype: float64

Tasa de cancelación
Dataset completo: 0.37041628277075134
Duplicados: 0.583866550479273


Los duplicados no representan una muestra típica del dataset. Se concentran principalmente en reservas del segmento Groups, tienen una presencia mucho mayor de depósitos Non Refund y presentan una tasa de cancelación significativamente más alta 58.4% vs 37%.

### 2.4 Criterio (7 pts)

El dataset no tiene identificador de reserva. Con base en la evidencia de 2.1, 2.2 y 2.3, responde:

a) ¿Las filas duplicadas son **errores de registro** o podrían ser **reservas distintas** que coinciden en todos sus valores? Argumenta con datos.

b) Las columnas `agent` y `company` tienen muchos faltantes. ¿Cómo trata `duplicated()` esos `NaN` y qué implicación tiene para este análisis?

A) Podrían ser reservas distintas. No hay evidencia suficiente para afirmar que son errores, ya que el dataset no tiene un identificador único.
Los duplicados se concentran en reservas de Groups 41.6%, tienen más Non Refund 35.6% y una mayor tasa de cancelación 58.4%, lo que sugiere un patrón específico y no errores aleatorios.

B)duplicated() considera iguales los NaN que están en la misma columna de dos filas.
Es decir si dos reservas tienen los mismos valores y además ambas tienen NaN en agent y company, serán consideradas duplicadas.

la implicacion es que como agent y company tienen muchos valores faltantes, esto puede aumentar la cantidad de filas detectadas como duplicadas.

**Respuesta 2.4:**

A) Podrían ser reservas distintas. No hay evidencia suficiente para afirmar que son errores, ya que el dataset no tiene un identificador único. Los duplicados se concentran en reservas de Groups 41.6%, tienen más Non Refund 35.6% y una mayor tasa de cancelación 58.4%, lo que sugiere un patrón específico y no errores aleatorios.

B)duplicated() considera iguales los NaN que están en la misma columna de dos filas. Es decir si dos reservas tienen los mismos valores y además ambas tienen NaN en agent y company, serán consideradas duplicadas.

la implicacion es que como agent y company tienen muchos valores faltantes, esto puede aumentar la cantidad de filas detectadas como duplicadas.

---
## Actividad 3. Impacto en el análisis (25 puntos)

### 3.1 Versión sin duplicados (5 pts)

Crea `hoteles_limpio` eliminando los duplicados exactos, conservando la primera aparición y **renumerando el índice**. Imprime el número de filas antes y después y cuántas se eliminaron.

In [8]:
hoteles_limpio = hoteles.drop_duplicates(keep='first').reset_index(drop=True)

# Cantidad de filas
filas_antes = len(hoteles)
filas_despues = len(hoteles_limpio)
filas_eliminadas = filas_antes - filas_despues

print(f"Filas antes de eliminar duplicados: {filas_antes}")
print(f"Filas después de eliminar duplicados: {filas_despues}")
print(f"Filas eliminadas: {filas_eliminadas}")

Filas antes de eliminar duplicados: 119390
Filas después de eliminar duplicados: 87396
Filas eliminadas: 31994


### 3.2 Métricas por hotel (8 pts)

Para `hoteles` y para `hoteles_limpio`, calcula por hotel con `groupby()`:

- Número de reservas (`.size()`).
- Tasa de cancelación (media de `is_canceled`).
- Tarifa diaria promedio (media de `adr`).

Nota: para calcular la media de varias columnas a la vez, selecciona una lista de columnas: `df.groupby('hotel')[['col1', 'col2']].mean()`.

Guarda las medias en dos `DataFrame`, `resumen_original` y `resumen_limpio`, con `hotel` como columna (usa `reset_index()`).

In [9]:
resumen_original = (
    hoteles.groupby('hotel')
    .agg(
        reservas=('hotel', 'size'),
        tasa_cancelacion=('is_canceled', 'mean'),
        adr_promedio=('adr', 'mean')
    )
    .reset_index()
)

# sin duplicados
resumen_limpio = (
    hoteles_limpio.groupby('hotel')
    .agg(
        reservas=('hotel', 'size'),
        tasa_cancelacion=('is_canceled', 'mean'),
        adr_promedio=('adr', 'mean')
    )
    .reset_index()
)

print("Resumen original")
print(resumen_original)

print("\nResumen limpio")
print(resumen_limpio)

Resumen original
          hotel  reservas  tasa_cancelacion  adr_promedio
0    City Hotel     79330          0.417270    105.304465
1  Resort Hotel     40060          0.277634     94.952930

Resumen limpio
          hotel  reservas  tasa_cancelacion  adr_promedio
0    City Hotel     53428          0.300386    110.985944
1  Resort Hotel     33968          0.234809     99.025346


### 3.3 Tabla comparativa (6 pts)

Agrega a cada resumen una columna `version` (`'original'` y `'sin duplicados'`) y únelos con `pd.concat()` en un `DataFrame` llamado `comparacion`, con índice renumerado y ordenado por `hotel`, de modo que las dos versiones de cada hotel queden juntas.

In [10]:

resumen_original['version'] = 'original'
resumen_limpio['version'] = 'sin duplicados'

comparacion = pd.concat(
    [resumen_original, resumen_limpio],
    ignore_index=True
)

comparacion = comparacion.sort_values(
    by=['hotel', 'version']
).reset_index(drop=True)

print(comparacion)

          hotel  reservas  tasa_cancelacion  adr_promedio         version
0    City Hotel     79330          0.417270    105.304465        original
1    City Hotel     53428          0.300386    110.985944  sin duplicados
2  Resort Hotel     40060          0.277634     94.952930        original
3  Resort Hotel     33968          0.234809     99.025346  sin duplicados


### 3.4 Interpretación (6 pts)

a) ¿Qué métrica cambia más al eliminar duplicados? ¿En qué hotel el cambio es mayor? Explica por qué, relacionándolo con lo encontrado en la Actividad 2.

b) Si se entrenara un modelo para predecir `is_canceled` con los datos originales, ¿qué problemas causarían los duplicados en el entrenamiento y en la evaluación del modelo?

**Respuesta 3.4:**

a) La tasa de cancelación es la métrica que más cambia, porque los registros duplicados tienen una tasa de cancelación mucho mayor 58.4% que el dataset completo 37.0%.

El cambio es mayor en el City Hotel, ya que concentra 25,902 duplicados 32.65%, mientras que el Resort Hotel tiene 6,092 duplicados 15.21%.

b) Se tendreia sesgo en el entrenamient ya que los registros repetidos tienen más peso y el modelo aprende de esos patrones.Tambien un sobreajuste ya que el modelo puede depender en exceso de características frecuentes en los duplicados, como reservas Groups y Non Refund.
otro problema podria ser la evaluación inflada ya que si las observaciones duplicadas aparecen tanto en entrenamiento como en prueba, las métricas serán  mejores.

Es decir los duplicados pueden hacer que el modelo parezca más preciso de lo que realmente es.

---
## Actividad 4. Tratamiento y verificación (20 puntos)

### 4.1 Verificación (5 pts)

Verifica sobre `hoteles_limpio` que:

- No quedan duplicados exactos.
- No hay índices repetidos y el índice va de 0 a `len(hoteles_limpio) - 1`.

In [11]:

duplicados_restantes = hoteles_limpio.duplicated().sum()

print("Duplicados restantes:", duplicados_restantes)

indice_correcto = (
    hoteles_limpio.index.is_unique and
    (hoteles_limpio.index.min() == 0) and
    (hoteles_limpio.index.max() == len(hoteles_limpio) - 1)
)

print("Indice correcto:", indice_correcto)


Duplicados restantes: 0
Indice correcto: True


### 4.2 Duplicados parciales (10 pts)

La columna `reservation_status_date` registra la fecha del último cambio de estado. Detecta las filas de `hoteles_limpio` que **coinciden en todas las columnas excepto `reservation_status_date`**.

1. Cuenta cuántas filas sobrantes hay con ese criterio.
2. Muestra algunas de esas filas ordenadas, con las columnas `hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adr`, `reservation_status` y `reservation_status_date`.
3. Muestra la distribución de `reservation_status` en esas filas.

Después responde: ¿qué podrían representar estas filas? ¿Las eliminarías? Justifica.

In [12]:

cols = [c for c in hoteles_limpio.columns
        if c != 'reservation_status_date']


dup_fecha = hoteles_limpio[
    hoteles_limpio.duplicated(subset=cols, keep='first')
]

print("Filas sobrantes con este criterio:", len(dup_fecha))

columnas_mostrar = [
    'hotel',
    'lead_time',
    'arrival_date_year',
    'arrival_date_month',
    'arrival_date_day_of_month',
    'adr',
    'reservation_status',
    'reservation_status_date'
]

ejemplos = hoteles_limpio[
    hoteles_limpio.duplicated(subset=cols, keep=False)
].sort_values(
    by=[
        'hotel',
        'lead_time',
        'arrival_date_year',
        'arrival_date_month',
        'arrival_date_day_of_month',
        'adr'
    ]
)

print(ejemplos[columnas_mostrar].head(20))


print("\nDistribución de reservation_status:")
print(dup_fecha['reservation_status'].value_counts(normalize=True))

Filas sobrantes con este criterio: 245
            hotel  lead_time  arrival_date_year arrival_date_month  \
35572  City Hotel          9               2015          September   
35574  City Hotel          9               2015          September   
51267  City Hotel         12               2017               July   
51268  City Hotel         12               2017               July   
54170  City Hotel         14               2015            October   
54474  City Hotel         14               2015            October   
39425  City Hotel         14               2016              April   
39432  City Hotel         14               2016              April   
40080  City Hotel         14               2016                May   
40083  City Hotel         14               2016                May   
45793  City Hotel         15               2016           December   
45797  City Hotel         15               2016           December   
48204  City Hotel         17               2017    

**Respuesta 4.2:**
Las filas probablemente representan la misma reserva con diferentes fechas de actualización del estado, ya que todos los datos son iguales y solo cambia reservation_status_date por lo que no las eliminaría. Aunque son muy parecidas, contienen información diferente en reservation_status_date, por lo que no son duplicados exactos. Eliminar una de ellas podría hacer perder información sobre cuándo cambió el estado de la reserva.

### 4.3 Registro de decisión (5 pts)

Redacta, en una celda de texto, el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Criterio** de duplicado utilizado (columnas comparadas).
- **Regla de conservación** (`keep`) y por qué es o no relevante en este caso.
- **Resultado** (filas antes, después y eliminadas).
- **Supuestos y limitaciones**.

**Respuesta 4.3:**
**Criterio**
 Se identificaron duplicados exactos comparando las 32 columnas del dataset, considerando como duplicadas aquellas filas que tenían los mismos valores en todas las variables registradas, incluyendo los valores faltantes (NaN) en las columnas agent y company.

**Regla de conservación:**
 Se utilizó keep='first', conservando la primera aparición de cada grupo de filas idénticas y eliminando las restantes.

**Resultado de la limpieza:**


*   Filas originales: 119,390
*   Filas después de eliminar duplicados: 87,396
*   Filas eliminadas: 31,994
*   Porcentaje eliminado: 26.80%


**Supuestos y limitaciones:**

El dataset no contiene un identificador único de reserva, por lo que no puede asegurarse que correspondan a errores de registro.
Algunas filas idénticas podrían representar reservas distintas con características exactamente iguales.
Los duplicados se concentran principalmente en reservas del segmento Groups, con depósitos Non Refund y alta tasa de cancelación, por lo que su eliminación podria afectar el análisis.


---
## Reflexión final (10 puntos)

Responde brevemente, con base en tus resultados:

1. Describe una situación en la que **no** convendría eliminar los duplicados exactos de este dataset.
2. ¿Qué información adicional (columnas) le pedirías al proveedor de los datos para tomar una decisión más segura? ¿Por qué?
3. En tus propias palabras, ¿por qué `duplicated().sum() == 0` no garantiza que un dataset esté libre de duplicados?

**Respuesta final:**

1. No convendría eliminarlos si varias reservas distintas pueden tener exactamente las mismas características y el dataset no incluye un identificador único. En este caso, eliminar los duplicados podría reducir el número real de reservas, especialmente en segmentos como Groups, donde observamos una alta concentración de registros idénticos.

2. Pediria:
ID único de reserva.
ID del cliente.
Fecha y hora de creación de la reserva.
Fecha y hora de las modificaciones realizadas.

Estas variables permitirían distinguir entre una reserva repetida por error y varias reservas legítimas con los mismos valores, facilitando una decisión sobre si conservar o eliminar los registros.

3. Porque duplicated() solo detecta filas exactamente iguales en las columnas comparadas. Dos registros pueden corresponder a la misma reserva y diferir en una sola variable (por ejemplo, reservation_status_date), por lo que no serían marcados como duplicados.

Es decir el resultado de 0 duplicados exactos no significa que no existan registros repetidos, sino únicamente que no hay filas completamente idénticas.

---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.